pinot-core/src/main/java/com/linkedin/pinot/core/periodictask/PeriodicTaskScheduler.java

```diff
@@ -0,0 +1,109 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.core.periodictask;
+
+import java.util.concurrent.Executors;
+import java.util.concurrent.PriorityBlockingQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class PeriodicTaskScheduler {
+  private static final Logger LOGGER = LoggerFactory.getLogger(PeriodicTaskScheduler.class);
+  private PriorityBlockingQueue<PeriodicTask> _periodicTasks;
+  private ScheduledExecutorService _executorService;
+  private long _intervalSeconds;
+  private long _initialDelaySeconds;
+
+  public PeriodicTaskScheduler() {
+    _executorService = Executors.newSingleThreadScheduledExecutor(runnable -> {
+      Thread thread = new Thread(runnable);
+      thread.setName("PeriodicTaskSchedulerExecutorService");
+      return thread;
+    });
+    _initialDelaySeconds = 120L;
+    _intervalSeconds = 300L;
+    _periodicTasks = new PriorityBlockingQueue<>(10, (o1, o2) -> {
+      if (o1.getExecutionTime() == o2.getExecutionTime()) {
+        return 0;
+      }
+      return o1.getExecutionTime() < o2.getExecutionTime() ? -1 : 1;
+    });
+  }
+
+  public void start() {
+    LOGGER.info("Starting PeriodicTaskScheduler. Run frequency in seconds: {}", _intervalSeconds);
+
+    // Set up an executor that executes tasks periodically
+    _executorService.schedule(() -> {
+      while (true) {
+        PeriodicTask task = null;
+        try {
+          task = _periodicTasks.take();
+          if (System.currentTimeMillis() < task.getExecutionTime()) {
+            Thread.sleep(System.currentTimeMillis() - task.getExecutionTime());
+          }
+          task.runTask();
+        } catch (InterruptedException ie) {
+          LOGGER.warn("Interrupted when running periodic task");
+          Thread.currentThread().interrupt();
+        } catch (Throwable e) {
+          // catch all errors to prevent subsequent executions from being silently suppressed
+          // Ref: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ScheduledExecutorService.html#scheduleWithFixedDelay-java.lang.Runnable-long-long-java.util.concurrent.TimeUnit-
+          LOGGER.warn("Caught exception while running PeriodicTaskScheduler", e);
+        } finally {
+          if (task != null) {
+            task.updateExecutionTime();
+            _periodicTasks.offer(task);
+          }
+        }
+      }
+    }, _initialDelaySeconds, TimeUnit.SECONDS);
+  }
+
+  public void stop() {
+    LOGGER.info("Stopping PeriodicTaskScheduler");
+    if (_executorService == null) {
+      return;
+    }
+    _executorService.shutdown();
```

Comment:
This will never terminate for an infinite loop, use **shutdownNow** instead

Reply:
@Jackie-Jiang Nope, as long as we set `_running` to false, it will terminate after finishing a task.
And using `shutdownNow` will throw a lot of unnecessary exceptions from either `take()` method or `sleep()` method.